sqlflow_models/gcn.py

```diff
@@ -0,0 +1,408 @@
+            return [loss_value, accuarcy]
+
+        return accuarcy
+
+    def sqlflow_train_loop(self, x):
+        """Customized training function."""
+        # load data
+        ids, ids_check, features, labels, edges, edge_check = list(), dict(), list(), list(), list(), dict()
+        from_node = 0
+        for inputs, label in x:
+            id = inputs[self.id_col].numpy().astype(np.int32)
+            feature = inputs[self.feature_col].numpy().astype(np.float32)
+            from_node = inputs[self.from_node_col].numpy().astype(np.int32)
+            to_node = inputs[self.to_node_col].numpy().astype(np.int32)
+            if int(id) not in ids_check:
+                ids.append(int(id))
+                features.append(feature)
+                labels.append(label.numpy()[0])
+                ids_check[int(id)] = 0
+            if tuple([int(from_node), int(to_node)]) not in edge_check:
+                edge_check[tuple([int(from_node), int(to_node)])] = 0
+                edges.append([from_node, to_node])
+        features = np.stack(features)
+        labels = np.stack(labels)
+        edges = np.stack(edges)
+
+        self.features, self.labels, self.adjacency, self.nshape, na_shape = self.preprocess(ids, features, labels, edges)
+        # training the model
+        wait = 0
+        best_acc = -9999999
+        PATIENCE = 10
+        for epoch in range(self.epochs):
+            # calculate the gradients and take the step
+            loss_value, grads = self.grad(self, [self.features, self.adjacency], self.labels, self.train_mask)
+            optimizer().apply_gradients(zip(grads, self.trainable_variables))
+            # Test on train and evaluate dataset
+            train_acc = self.test(self.train_mask)
+            val_acc = self.test(self.val_mask)
+            print("Epoch {} loss={:6f} accuracy={:6f} val_acc={:6f}".format(epoch, loss_value, train_acc, val_acc))
+            # early stopping
+            if epoch > 50 and self.early_stopping:
+                if float(val_acc.numpy()) > best_acc:
+                    best_acc = float(val_acc.numpy())
+                    wait = 0
+                else:
+                    if wait >= PATIENCE:
+                        print('Epoch {}: early stopping'.format(epoch))
+                        break
+                    wait += 1
+        # evaluate the model
+        result = self.evaluate(data=[self.features, self.adjacency], y=self.labels, sample_weight=self.val_mask)
+        # get all the results
+        predicted = self.predict([self.features, self.adjacency])
+        # store the results in a pickled file
+        with open('./results.pkl', 'wb') as f:
+            results = dict()
+            for i in range(len(ids)):
+                results[str(ids[i])] = predicted[i]
+            results['evaluation'] = result
+            pickle.dump(results, f)
+            self.results = results
+
+    def sqlflow_evaluate_loop(self, x, metric_names):
```

Comment:
I guess we still need it if we want to have `EVALUATE` and `PREDICT` commands through sqlflow. Otherwise, we have to print out the results for both evaluation and prediction during training phase. (This seems to be a bad idea for me.) What do other people think? @typhoonzero, @lhw362950217 (Ideas are greatly appreciated!)

Reply:
I think we may keep it to adapt the `PREDICT` and `EVALUATE` statements.

Reply:
I agree with @Derek-Wds .